packages/vite/src/node/ssr/ssrExternal.ts

```diff
@@ -175,9 +175,26 @@ function collectExternals(
 
 export function shouldExternalizeForSSR(
   id: string,
-  externals: string[]
+  externals: string[],
+  noExternals: string | RegExp | (string | RegExp)[] | true
 ): boolean {
+  const noExternalsArr: RegExp[] =
+    noExternals !== true
+      ? (Array.isArray(noExternals) ? noExternals : [noExternals]).map(
+          (item) => {
+            if (typeof item === 'string') {
+              item = new RegExp(item)
+            }
+            return item
+          }
+        )
+      : []
   const should = externals.some((e) => {
+    for (const r of noExternalsArr) {
+      if (r.test(id)) {
+        return false
+      }
+    }
```

Comment:
**suggestion (non-blocking):** This doesn't access `e`, so we could move it into a `filter`

Reply:
Move it out of loop 🤔